python/lsst/pipe/tasks/processCcdWithFakes.py

```diff
@@ -309,6 +326,111 @@ def run(self, fakeCat, exposure, wcs=None, photoCalib=None, exposureIdInfo=None)
         self.measurement.run(measCat=sourceCat, exposure=exposure, exposureId=exposureIdInfo.expId)
         self.applyApCorr.run(catalog=sourceCat, apCorrMap=exposure.getInfo().getApCorrMap())
         self.catalogCalculation.run(sourceCat)
+        sourceCat = self.copyIcSourceFields(icSourceCat, sourceCat)
 
         resultStruct = pipeBase.Struct(outputExposure=exposure, outputCat=sourceCat)
         return resultStruct
+
+    def copyIcSourceFields(self, icSourceCat, sourceCat):
+        """Match sources in icSourceCat and sourceCat and copy the specified fields
+
+        Parameters
+        ----------
+        icSourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog from which to copy fields.
+        sourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog to which to copy fields.
+
+        Returns
+        -------
+        newCat : `lsst.afw.table.SourceCatalog`
+            Catalog which includes the copied fields.
+
+        The fields copied are those specified by `config.icSourceFieldsToCopy` that actually exist
+        in the schema of `icSourceCat`.
+
+        This version was based on and adapted from the one in calibrateTask.
+        """
+
+        # Make a new SourceCatalog with the data from sourceCat so that we can add the new columns to it
+        sourceSchemaMapper = afwTable.SchemaMapper(sourceCat.schema)
+        sourceSchemaMapper.addMinimalSchema(sourceCat.schema, True)
+
+        icSourceSchemaMapper = afwTable.SchemaMapper(icSourceCat.schema, sourceCat.schema)
+
+        # Add the desired columns from the config option icSourceFieldsToCopy
+        missingFieldNames = []
+        for fieldName in self.config.icSourceFieldsToCopy:
+            try:
+                schemaItem = icSourceCat.schema.find(fieldName)
+            except KeyError:
+                missingFieldNames.append(fieldName)
+            else:
+                # field found; if addMapping fails then raise an exception
+                icSourceSchemaMapper.editOutputSchema().addField(schemaItem.getField())
+                schema = icSourceSchemaMapper.editOutputSchema()
+                icSourceSchemaMapper.addMapping(schemaItem.getKey(), schema.find(fieldName).getField())
+
+        if missingFieldNames:
+            raise RuntimeError("isSourceCat is missing fields {} specified in "
+                               "icSourceFieldsToCopy".format(missingFieldNames))
+
+        self.calibSourceKey = icSourceSchemaMapper.addOutputField(afwTable.Field["Flag"]("calib_detected",
+                                                                  "Source was detected as an icSource"))
+
+        schema = icSourceSchemaMapper.getOutputSchema()
+        newCat = afwTable.SourceCatalog(schema)
+        newCat.reserve(len(sourceCat))
```

Comment:
It's prolly not necessary (I think `extend` does it for you), but glad to see you're thinking about memory allocation issues.